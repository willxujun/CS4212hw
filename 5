/*
Source: Paul Tan Han Kiat
005.pass.in
 * Testing associativity / precedence and pretty printing.
 */

class Main {

    Void main() {

        // || is left associative
        return 4 <= 2 || 9 != 8 || false;

        // && is left associative
        return null && 10 < 9 && foo("bar");

        // + is left associative
        return 4 + 3 + 2;

        // - is left associative
        return 10 - 9 - 8;

        // * is left associative
        return 10 * 3 * 99;

        // / is left associative
        return 99 / 2 / 3;

        // || has greater precedence than &&
        return true || false && true || false;
        return true && false || true && false;

        // atoms have the highest precedence
        return !new Bar();
        return !foo("bar");

        // Unneeded brackets will be removed by the pretty-printer
        return (4 * 3) + (9 / 2);

        // atoms have the highest precedence
        if (4 >= 2 && 3 < new Bar() || foo("bar") && 3 / 2 == 6 && null || false) {
            // + and - have the same precedence
            // * and / have the same precedence, and have greater precedence than +/-
            return 1 + 2 - 3 - 8 + 4 * 8 / 3 * 2 / 9 + 2;
        } else {
            return !(false && !true);
        }

        return 3 < 2 + 8 || 4 >= 6 * 3;

        // Left associative
        a.b.c.d.e.f.g(h.i.j.k);

        // Left associative
        a.b.c.d = e;
    }

}
