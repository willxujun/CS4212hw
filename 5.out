instead expected token classes are [LPAREN, DOT, ASSIGN]
instead expected token classes are [RBRACES, TYPEID]
Parse error, or failed distinct name check. See parse tree for detail.
Program
  Error: Syntax error in class defn
  Error: Syntax error in class defn
  Class OK
    VariableList
      Int a
      String b
    MethodList
      Bool do
        FormalList
          Int t1
          Int t2
          Int t3
          Int t4
        VariableList
        StatementList
          Return
            (BinExp And)
              BinExp And
                UExp Not
                  (BinExp And)
                    BinExp And
                      (BinExp Not)
                        BinExp Not
                          (Id t1)
                          (Id t2)
                      (Id t3)
                (Id t4)

