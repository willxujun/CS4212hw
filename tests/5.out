Error: l expression does not have bool or int type in (Id foo) NE
In method main
In class Main

Error: l expression does not have bool or int type in BinExp NE And
In method main
In class Main

Error: l expression does not have bool or int type in BinExp And And
In method main
In class Main

Error: While statement has non-bool test type Error_t
In method main
In class Main

Error: l expression does not have bool or int type in (Id foo) NE
In method main
In class Main

Error: l expression does not have bool or int type in BinExp NE And
In method main
In class Main

Error: l expression does not have bool or int type in BinExp And Or
In method main
In class Main

Error: If statement has non-bool test type Error_t
In method main
In class Main

Program
  Class Main
    VariableList
    MethodList
      Void main
        FormalList
        VariableList
          Int k
          Bool bool1
          Foo foo
        StatementList
          Assign foo
            New Foo
          While
            BinExp And
              BinExp And
                BinExp NE
                  (Id foo)
                  Null
                (Id bool1)
              Dispatch
                Access method1
                  (Id foo)
                ExpressionList
            StatementList
              Assign k
                (Int 1)
          Assign k
            BinExp Plus
              BinExp Div
                (Int 2)
                (Int 1)
              BinExp Mult
                (Int 4)
                (Int 5)
          Assign bool1
            BinExp Or
              (Bool true)
              Dispatch
                Access method1
                  (Id foo)
                ExpressionList
          Assign bool
            New Foo
            BinExp Or
              BinExp Or
                BinExp Or
                  BinExp GE
                    (Int 4)
                    (Int 2)
                  (Bool true)
                BinExp LT
                  (Int 4)
                  (Int 5)
              (Bool false)
          If
            BinExp Or
              BinExp And
                BinExp NE
                  (Id foo)
                  Null
                Dispatch
                  Access method1
                    (Id foo)
                  ExpressionList
              UExp Not
                Access bool
                  (Id foo)
            StatementList
              Assign k
                (Int 0)
            StatementList
              Assign k
                (Int 1)
          While
            BinExp Or
              (Id bool1)
              BinExp And
                BinExp GT
                  (Int 4)
                  (Int 5)
                UExp Not
                  Dispatch
                    Access method1
                      (Id foo)
                    ExpressionList
            StatementList
              Assign k
                (Int 2)
  Class Foo
    VariableList
      Bool bool
    MethodList
      Bool method1
        FormalList
        VariableList
        StatementList
          Return
            BinExp GT
              (Int 1)
              (Int 2)
      Void method2
        FormalList
        VariableList
        StatementList
          Print
            (Str (Hello!))

Semantics error detected. Refer to msgs and parse tree.
